File: problema1/pascal.py
import time

def generar_pascal(n):
    """
    Parte A: Genera los coeficientes usando el Triángulo de Pascal.
    En Python, las listas (arrays) manejan la memoria dinámica automáticamente.
    """
    # Inicializamos dinámicamente la primera fila
    fila_anterior = [1]

    for i in range(1, n + 1):
        # Reservamos dinámicamente espacio para la nueva fila
        fila_actual = [1] * (i + 1)
        
        # Calculamos los valores intermedios sumando la fila anterior
        for j in range(1, i):
            fila_actual[j] = fila_anterior[j - 1] + fila_anterior[j]
        
        # Asignamos la nueva fila.
        # El recolector de basura de Python libera la memoria de la fila vieja.
        fila_anterior = fila_actual

    return fila_anterior

def mostrar_polinomio(coeficientes, n):
    """
    Muestra el polinomio formateado estéticamente.
    """
    terminos = []
    
    for i, coef in enumerate(coeficientes):
        potencia = n - i
        
        if potencia == 0:
            terminos.append(str(coef))
        elif potencia == 1:
            terminos.append(f"{'' if coef == 1 else coef}x")
        else:
            terminos.append(f"{'' if coef == 1 else coef}x^{potencia}")
            
    print("\n======================================================")
    print(f"   a) POLINOMIO GENERADO PARA (x + 1)^{n}")
    print("======================================================")
    print(f"f(x) = {' + '.join(terminos)}")
    print("------------------------------------------------------")

def evaluar_paso_a_paso(coeficientes, n, x_valor):
    """
    Evalúa f(x) mostrando la resolución natural paso a paso (Parte B).
    """
    print("\n======================================================")
    print(f"   b) EVALUACIÓN PASO A PASO PARA x = {x_valor}")
    print("======================================================")
    
    paso1_sustitucion = []
    paso2_potencias = []
    paso3_multiplicar = []
    suma_total = 0

    for i, coef in enumerate(coeficientes):
        potencia = n - i
        
        # 1. Valores para Paso 1: Sustitución explícita
        if potencia == 0:
            paso1_sustitucion.append(f"{coef}")
        elif potencia == 1:
            paso1_sustitucion.append(f"{coef}({x_valor})")
        else:
            paso1_sustitucion.append(f"{coef}({x_valor})^{potencia}")

        # 2. Valores para Paso 2: Resolver las potencias
        valor_potencia = x_valor ** potencia
        if potencia == 0:
            paso2_potencias.append(f"{coef}")
        else:
            paso2_potencias.append(f"{coef}({valor_potencia})")

        # 3. Valores para Paso 3 y 4: Multiplicaciones y Suma
        resultado_termino = coef * valor_potencia
        paso3_multiplicar.append(str(resultado_termino))
        suma_total += resultado_termino

    # Imprimir el flujo completo en bloques ordenados
    print(f"Paso 1 (Sustitución) : f({x_valor}) = {' + '.join(paso1_sustitucion)}")
    print(f"Paso 2 (Potencias)   : f({x_valor}) = {' + '.join(paso2_potencias)}")
    print(f"Paso 3 (Multiplicar) : f({x_valor}) = {' + '.join(paso3_multiplicar)}")
    print("------------------------------------------------------")
    print(f"Paso 4 (Resultado)   : f({x_valor}) = {suma_total}")
    print("======================================================\n")

def medir_tiempo_n100():
    """
    Función requerida para medir el tiempo de n=100 y escribirlo en el txt.
    """
    inicio = time.perf_counter()
    generar_pascal(100)
    fin = time.perf_counter()
    
    segundos = fin - inicio
    texto_resultado = f"Python - Tiempo de ejecucion para generar Pascal (n=100): {segundos:.6f} segundos\n"

    try:
        with open("tiempos_ejecucion.txt", "a") as archivo:
            archivo.write(texto_resultado)
        print(f"[INFO] Prueba de estrés completada. Resultado guardado en tiempos_ejecucion.txt: {segundos:.6f} seg.")
    except Exception as e:
        print("Error al escribir el archivo txt:", e)

def main():
    try:
        n = int(input("Ingrese el exponente n (entero no negativo): "))
        if n < 0:
            print("Error: Debe ingresar un número entero mayor o igual a cero.")
            return

        x = int(input("Ingrese el valor entero de x para evaluar el polinomio: "))
        
        # Ejecutamos Parte A midiendo el tiempo exacto
        inicio = time.perf_counter()
        coeficientes = generar_pascal(n)
        fin = time.perf_counter()
        segundos = fin - inicio
        
        mostrar_polinomio(coeficientes, n)

        # Ejecutamos Parte B
        evaluar_paso_a_paso(coeficientes, n, x)

        # Escribimos el tiempo real en el archivo txt
        texto_resultado = f"Python - Tiempo para generar Pascal (n={n}): {segundos:.6f} segundos\n"
        try:
            with open("tiempos_ejecucion.txt", "a") as archivo:
                archivo.write(texto_resultado)
            print(f"[INFO] Tiempo de cálculo guardado en tiempos_ejecucion.txt: {segundos:.6f} seg.")
        except Exception as e:
            print("Error al escribir el archivo txt:", e)
        
    except ValueError:
        print("Error: Entrada inválida. Por favor ingrese números enteros.")

if __name__ == "__main__":
    main()

File: problema1/pascal.js
const fs = require('fs');
const { performance } = require('perf_hooks');
const readline = require('readline');

// Configuración para leer datos desde la consola
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

/**
 * Parte A: Genera los coeficientes usando el Triángulo de Pascal.
 * Utiliza arreglos dinámicos y el recolector de basura nativo.
 */
function generarPascal(n) {
    // Inicializamos dinámicamente la primera fila
    let filaAnterior = [1n];

    for (let i = 1; i <= n; i++) {
        // Reservamos un nuevo arreglo dinámico para la fila actual
        let filaActual = new Array(i + 1);
        filaActual[0] = 1n; // Extremo izquierdo
        filaActual[i] = 1n; // Extremo derecho

        // Calculamos los valores intermedios sumando la fila anterior
        for (let j = 1; j < i; j++) {
            filaActual[j] = filaAnterior[j - 1] + filaAnterior[j];
        }

        // Asignamos la nueva fila. 
        // El recolector de basura libera automáticamente la memoria de la fila anterior.
        filaAnterior = filaActual;
    }

    return filaAnterior;
}

/**
 * Muestra el polinomio formateado estéticamente (Parte A)
 */
function mostrarPolinomio(coeficientes, n) {
    let terminos = [];
    
    for (let i = 0; i < coeficientes.length; i++) {
        let potencia = n - i;
        let coef = coeficientes[i];

        if (potencia === 0) {
            terminos.push(coef.toString());
        } else if (potencia === 1) {
            terminos.push(`${coef === 1n ? "" : coef}x`);
        } else {
            terminos.push(`${coef === 1n ? "" : coef}x^${potencia}`);
        }
    }
    
    console.log(`\n======================================================`);
    console.log(`   a) POLINOMIO GENERADO PARA (x + 1)^${n}`);
    console.log(`======================================================`);
    console.log(`f(x) = ${terminos.join(" + ")}`);
    console.log(`------------------------------------------------------`);
}

/**
 * Evalúa f(x) mostrando la resolución natural paso a paso (Parte B)
 */
function evaluarPasoAPaso(coeficientes, n, xValor) {
    console.log(`\n======================================================`);
    console.log(`   b) EVALUACIÓN PASO A PASO PARA x = ${xValor}`);
    console.log(`======================================================`);
    
    let x = BigInt(xValor);
    
    let paso1Sustitucion = [];
    let paso2Potencias = [];
    let paso3Multiplicar = [];
    let sumaTotal = 0n;

    for (let i = 0; i < coeficientes.length; i++) {
        let potencia = BigInt(n - i);
        let coef = coeficientes[i];
        
        // 1. Valores para Paso 1: Sustitución explícita
        if (potencia === 0n) {
            paso1Sustitucion.push(`${coef}`);
        } else if (potencia === 1n) {
            paso1Sustitucion.push(`${coef}(${xValor})`);
        } else {
            paso1Sustitucion.push(`${coef}(${xValor})^${potencia}`);
        }

        // 2. Valores para Paso 2: Resolver las potencias
        let valorPotencia = x ** potencia;
        if (potencia === 0n) {
            paso2Potencias.push(`${coef}`);
        } else {
            paso2Potencias.push(`${coef}(${valorPotencia})`);
        }

        // 3. Valores para Paso 3 y 4: Multiplicaciones y Suma
        let resultadoTermino = coef * valorPotencia;
        paso3Multiplicar.push(`${resultadoTermino}`);
        sumaTotal += resultadoTermino;
    }

    // Imprimir el flujo completo en bloques ordenados
    console.log(`Paso 1 (Sustitución) : f(${xValor}) = ${paso1Sustitucion.join(" + ")}`);
    console.log(`Paso 2 (Potencias)   : f(${xValor}) = ${paso2Potencias.join(" + ")}`);
    console.log(`Paso 3 (Multiplicar) : f(${xValor}) = ${paso3Multiplicar.join(" + ")}`);
    console.log(`------------------------------------------------------`);
    console.log(`Paso 4 (Resultado)   : f(${xValor}) = ${sumaTotal}`);
    console.log(`======================================================\n`);
}

/**
 * Función requerida para medir el tiempo de n=100 y escribirlo en el txt
 */
function medirTiempoN100() {
    const inicio = performance.now();
    generarPascal(100);
    const fin = performance.now();
    
    // Tiempo en milisegundos convertido a segundos
    const segundos = (fin - inicio) / 1000;
    const textoResultado = `JavaScript (Node.js) - Tiempo de ejecucion para generar Pascal (n=100): ${segundos.toFixed(6)} segundos\n`;

    try {
        fs.appendFileSync('tiempos_ejecucion.txt', textoResultado);
        console.log(`[INFO] Prueba de estrés completada. Resultado guardado en tiempos_ejecucion.txt: ${segundos.toFixed(6)} seg.`);
    } catch (error) {
        console.error("Error al escribir el archivo txt:", error);
    }
}

// Flujo principal del programa
function main() {
    rl.question("Ingrese el exponente n (entero no negativo): ", (respuestaN) => {
        let n = parseInt(respuestaN);
        
        if (isNaN(n) || n < 0) {
            console.log("Error: Debe ingresar un número entero mayor o igual a cero.");
            rl.close();
            return;
        }

        rl.question("Ingrese el valor entero de x para evaluar el polinomio: ", (respuestaX) => {
            let x = parseInt(respuestaX);

            if (isNaN(x)) {
                console.log("Error: Valor de x inválido.");
                rl.close();
                return;
            }

            // Ejecutamos Parte A midiendo el tiempo exacto
            const inicio = performance.now();
            let coeficientes = generarPascal(n);
            const fin = performance.now();
            const segundos = (fin - inicio) / 1000;

            mostrarPolinomio(coeficientes, n);

            // Ejecutamos Parte B
            evaluarPasoAPaso(coeficientes, n, x);

            // Escribimos el tiempo real en el archivo txt
            const textoResultado = `JavaScript (Node.js) - Tiempo para generar Pascal (n=${n}): ${segundos.toFixed(6)} segundos\n`;
            try {
                fs.appendFileSync('tiempos_ejecucion.txt', textoResultado);
                console.log(`[INFO] Tiempo de cálculo guardado en tiempos_ejecucion.txt: ${segundos.toFixed(6)} seg.`);
            } catch (error) {
                console.error("Error al escribir el archivo txt:", error);
            }

            rl.close();
        });
    });
}

// Iniciar programa
main();